### PR TITLE
Introduce IUIAutomation5 interface and ability to announce notifications in Windows 10 Fall Creators Update and later

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
#7984, #8009 

### Summary of the issue:
Introduce IUIAutomation5 interface and ability to announce notifications in Windows 10 Fall Creators Update an later

### Description of how this pull request fixes the issue:
In Windows 10 Fall Creators Update and later, apps can inform screen readers to announce specific notification text. This is done via IUIAutomationNotificationEventHandler::HandleNotificationEvent, which is part of IUIAutomation5 interface. Since NVDA only uses IUIAutomation3, update UIA subsystem to use IUIAutomation5. Also, introduce ability to use more recent IUIAutomation interfaces on various Windows 10 releases (such as IUIAutomation4 in Anniversary Update/build 14393).

### Testing performed:
Testing done via Windows 10 App Essentials add-on: using recent versions of various app (such as Calculator, Microsoft Store and others) that uses this new event to announce texts.

### Known issues with pull request:
None so far.

### Change log entry:
New feature: In Windows 10 Fall Creators Update and later, NVDA can announce notifications from apps such as Calculator and Windows Store.

### CODE REVIEW

**Comment by leonardder** [comment]
Hey Joseph,

1. It seems you defined the getIUIAInterface function, but you never use it in your code below.
2. As noted on Gitter, I'm honestly not a fan of the build version checking, as it does not work when running from source. Furthermore, I'm not convinced of whether it is significantly faster than just querying interfaces from 5 up to 2. Some timing on my end revealed that querying interfaces is actually very fast.

**Comment by michaelDCurran** [comment]
Querying interfaces on the CUIAutomation client object should be very 
fast as the UI automation client COM interfaces run entirely within 
NVDA's process.
Also, this will only be done once, on NVDA start anyway I'm assuming.

**Comment by michaelDCurran** [comment]
Do you know if this causes any double speaking where Microsoft may have fired both liveRegionChange and notification for the same thing?

**Comment by josephsl** [comment]
Hi, yes, but apps I came across so far raises one of these at runtime, or prefers the newer events. For example, I expected Microsoft Store app to fire either live region change or notification event in Creators Update (Version 1703), but it raised only notification event in that version. But I would imagine third-party apps may raise both events (perhaps due to oversight), but I guess preparing for that might be in order somehow. Thanks.

**Comment by josephsl** [comment]
Hi,

@LeonardDer: the Win10 version is the one to be used from binary builds and eventually from source once we move to Python 3. The generic version is meant for add-ons and to show that NVDA still supports Windows 7 SP1. The reason for this design is to keep it compatible with what we have at the moment: falling back to IUIAutomation2 if newer versions are not available. As @MichaelDCurran pointed out, this is meant to be used at startup (both for Core and add-ons).

Thanks.

**Comment by michaelDCurran** [comment]
As all IUIAutomationX interfaces inherit from the one before, there is 
no issue with compatibility. I.e. code written for windows 7 can 
interact with an IUIAutomation5 object just as it was an IUIAutomation2 
object.
At start-up, NVDA should simply QueryInterface from IUIAutomation5 down 
to IUIAutomation until one succeeds.
Thus UIAHandler.handler.clientObject will always be the highest one the 
Operating System happens to support.

**Comment by josephsl** [comment]
Hi,

In the end, the getIUIAInterface function won't be added, as a for loop will be used to try various versions (modifying specs accordingly).

Thanks.

File: source/NVDAObjects/UIA/__init__.py [hunk 1]
```diff
@@ -1368,6 +1369,16 @@ def event_UIA_systemAlert(self):
 		# Ideally, we wouldn't use getBrailleTextForProperties directly.
 		braille.handler.message(braille.getBrailleTextForProperties(name=self.name, role=self.role))
 
+	def event_UIA_notification(self, sender=None, notificationKind=None, notificationProcessing=None, displayString=None, activityId=None):
```
Comment (leonardder) [inline source/NVDAObjects/UIA/__init__.py]: I think you could leave out the sender kwarg. See my comment below.

File: source/_UIAHandler.py [hunk 2]
```diff
@@ -288,6 +296,14 @@ def IUIAutomationPropertyChangedEventHandler_HandlePropertyChangedEvent(self,sen
 			obj=focus
 		eventHandler.queueEvent(NVDAEventName,obj)
 
+	def IUIAutomationNotificationEventHandler_HandleNotificationEvent(self,sender,NotificationKind,NotificationProcessing,displayString,activityId):
+		if not self.MTAThreadInitEvent.isSet():
+			# UIAHandler hasn't finished initialising yet, so just ignore this event.
+			return
+		import NVDAObjects.UIA
+		obj=NVDAObjects.UIA.UIA(UIAElement=sender)
+		eventHandler.queueEvent("UIA_notification",obj, sender=sender, notificationKind=NotificationKind, notificationProcessing=NotificationProcessing, displayString=displayString, activityId=activityId)
```
Comment (leonardder) [inline source/_UIAHandler.py]: As the sender is equal to obj.UIAElement, I think it is a redundant kwarg. For consistency reasons, if one whould like to access the UIAElement for this event within a subclass, people can use the UIAElement property on the object.

File: source/_UIAHandler.py [hunk 3]
```diff
@@ -182,11 +182,16 @@ def MTAThreadFunc(self):
 			except (COMError,WindowsError,NameError):
 				self.clientObject=CoCreateInstance(CUIAutomation._reg_clsid_,interface=IUIAutomation,clsctx=CLSCTX_INPROC_SERVER)
 			if isUIA8:
-				try:
-					self.clientObject=self.clientObject.QueryInterface(IUIAutomation3)
-				except COMError:
-					self.clientObject=self.clientObject.QueryInterface(IUIAutomation2)
-			log.info("UIAutomation: %s"%self.clientObject.__class__.__mro__[1].__name__)
+				# #8009: use appropriate interface based on highest supported interface.
+				for interface in (IUIAutomation5, IUIAutomation4, IUIAutomation3, IUIAutomation2):
```
Comment (leonardder) [inline source/_UIAHandler.py]: May be put these interfaces in a constant?

File: source/_UIAHandler.py [hunk 4]
```diff
@@ -182,11 +182,16 @@ def MTAThreadFunc(self):
...
-			log.info("UIAutomation: %s"%self.clientObject.__class__.__mro__[1].__name__)
+				# #8009: use appropriate interface based on highest supported interface.
+				for interface in (IUIAutomation5, IUIAutomation4, IUIAutomation3, IUIAutomation2):
+					try:
+						self.clientObject=self.clientObject.QueryInterface(interface)
+						break
+					except COMError:
+						pass
+			# Have this handy because we want to add things corresponding to the interface version.
+			IUIAVersion = self.clientObject.__class__.__mro__[1].__name__
```
Comment (leonardder) [inline source/_UIAHandler.py]: Now you loop through these interfaces (which I think is a good thing to do), you might as well do something like this, avoiding digging into the mro:

```
for interface in interfaces:
    QueryInterface
except:
    pass # Or continue? I guess they are identical in result here
else:
    IUIAVersion = interface.__name__
    ...
```

File: source/_UIAHandler.py [hunk 5]
```diff
@@ -206,6 +211,9 @@ def MTAThreadFunc(self):
 			self.clientObject.AddPropertyChangedEventHandler(self.rootElement,TreeScope_Subtree,self.baseCacheRequest,self,UIAPropertyIdsToNVDAEventNames.keys())
 			for x in UIAEventIdsToNVDAEventNames.iterkeys():  
 				self.clientObject.addAutomationEventHandler(x,self.rootElement,TreeScope_Subtree,self.baseCacheRequest,self)
+			# #7984: add support for notification event (IUIAutomation5, part of Windows 10 build 16299 and later).
+			if IUIAVersion >= "IUIAutomation5":
```
Comment (leonardder) [inline source/_UIAHandler.py]: Hmm, even though Python uses lexicographical comparison for strings and therefore this is code that does the write thing, I'd feel safer if you'd avoid this. As @michaeldcurran pointed out in https://github.com/nvaccess/nvda/pull/8045#issuecomment-369188217, all UIA interfaces inherit from the one before, so you could do an instance check on the client object for IUIAutomation5.

**Review by LeonarddeR — COMMENTED** [review]
I know you didn't ask for my review, but I'm trying to update my UIA knowledge, so that's why I looked into this.

File: source/_UIAHandler.py [hunk 6]
```diff
@@ -182,11 +182,16 @@ def MTAThreadFunc(self):
...
-				except COMError:
-					self.clientObject=self.clientObject.QueryInterface(IUIAutomation2)
-			log.info("UIAutomation: %s"%self.clientObject.__class__.__mro__[1].__name__)
+				# #8009: use appropriate interface based on highest supported interface.
+				for interface in (IUIAutomation5, IUIAutomation4, IUIAutomation3, IUIAutomation2):
+					try:
+						self.clientObject=self.clientObject.QueryInterface(interface)
+						break
+					except COMError:
+						pass
```
Comment (michaelDCurran) [inline source/_UIAHandler.py]: Perhaps an else on the for loop that raises RuntimeError("No UIAutomation client interface supported")

File: source/_UIAHandler.py (hunk 4 above)
Comment (michaelDCurran) [reply source/_UIAHandler.py]: UiaVersion would also have to be set to IUIAutomation when isUIA8 is false. if you do it the way @leonardder suggests

**Comment by josephsl** [comment]
Hi,

In the end, isinstance checking will be done. I don't think an "else" is necessary except if we need to break out of the loop with that method.

Additional testing done by running the launcher (based on source code commits) on a Windows 8.1 system, and things work as expected. Thanks.

**Comment by Brian1Gaff** [comment]
Does this mean that if Microsoft bring out an even higher number this will 
need to be added or are they now finished twiddling about with the versions?

**Comment by josephsl** [comment]
Hi,

If the new interface includes useful features that screen reader users need to see/hear, then yes, we need to add support for it.

Thanks.

File: source/_UIAHandler.py (hunk 2 above)
Comment (michaelDCurran) [reply source/_UIAHandler.py]: I agree with @leonardder here. Sender should not be explicitly provided to the notification event as it is redundant.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

**Review by michaelDCurran — APPROVED** [review]

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**